### ISSUE TITLE
Bug: @typescript-eslint/prefer-optional-chain flags situations where an optional chain would change behavior

### ISSUE THREAD

**Issue by aboks** [typescript-eslint/typescript-eslint#11700]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.9.3&fileType=.tsx&code=CYUwxgNghgTiAEYD2A7AzgF3lAXPA3vAEYD8emMAligObwC%2BA3AFDOUBm8AFAIRTwAfAdgB0ReAF4p8AETskSGQEoCzePGb0gA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0WloHsalfkwCG8WmQAWo5uigB3UdCaRwYAL4gNQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
declare const a: { b?: string };

if (!a || a.b === "foo") {
  
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/prefer-optional-chain": ["error", {}],
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {}
}
```

### Expected Result

I expect that no error from `@typescript-eslint/prefer-optional-chain` is triggered on the third line, as an optional chain would change behavior. 

### Actual Result

An error from `@typescript-eslint/prefer-optional-chain` is triggered on the third line. When autofixed, this produces the code
```ts
declare const a: { b?: string };

if (a?.b === "foo") {
  
}
```
When `a.b` is undefined, the original comparison produced the value `true`, executing the code within the `if`-block. The 'autofixed' code produces `false` in that case, skipping the code within the `if`-block.

### Additional Info

This seems due to the changes in https://github.com/typescript-eslint/typescript-eslint/pull/11533. The comment in https://github.com/typescript-eslint/typescript-eslint/pull/11533#issuecomment-3400575571 already hints at this incorrect behavior.

**Comment by kirkwaiblinger** [comment]
Thanks for the report! Marked accepting PRs.

I notice that this example is a bit curious in that testing `!a` is an unnecessary condition, since `a` is not nullable. Have you also encountered a repro case where the code doesn't have the "unnecessary" nullish check? 

It's definitely a bug, of course, but I'm interested to know the extent of the reproduction conditions, if you have encountered other cases where the detection is over eager.

**Comment by kirkwaiblinger** [comment]
Cc @mdm317

**Comment by mdm317** [comment]
Apologies for the error in my PR. 


-----

> When a.b is undefined, the original comparison produced the value true

I might be missing something, but the original comparison actually produces `false` when a.b is undefined, right?
```
const a = {};
if( !a || a.b === "foo" )
     ^^^^^^^^^^^^^   false, right?
```
Happy to be corrected if I got this wrong.

-----


I’ve found that it’s a bug when the left operand type is a Record.
I think there might be more cases that produce the same bug, but I’ll need some time to check.
```
declare const record: Record<string, { kind: string }>;
// Should not trigger a lint error, but it does
if (record['key'] && record['key'].kind !== '1') {
}

```
[repro link](https://typescript-eslint.io/play/#ts=5.9.3&showAST=types&fileType=.tsx&code=CYUwxgNghgTiAEYD2A7AzgF3nZNgC54AlcJPAHkxgEsUBzAGngG94BrWg%2BK2u%2BAXwB8AbgBQAenHwAygAskAVwjB4KJFgw06dEDHhR4EWll0wyTAEYKs1LMCQg0o6gDN4AChxlgAbQDkbCAAnn4AuvAAZBHYpHj%2BgSGhAHQcKCoAhAC8mfB%2BAIx%2BAJQsovyioqCQsAjI6FgWhKxghDz0AvAAPqpKEGKSMvJKKmoaWjp6BkYoJjBmMExQafC28PaOKICYBBjObu4WkdEWSWDwWTn5RSX8QA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0WloHsalfkwCG8WmQAWo5uigB3UdCaRwYAL4gNQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

**Comment by Hypnosphi** [comment]
Same with arrays: https://typescript-eslint.io/play/#ts=5.9.3&fileType=.tsx&code=CYUwxgNghgTiAEYD2A7AzgF3rGUCeAXPAN7wBGA-EZjAJYoDm8AvgNoC6A3AFDe0Bm8ABQBCHPlYBGdvAA%2Bs7DFx4p7AHRl4AXh3wARPyRI9AShLd48bsyA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0WloHsalfkwCG8WmQAWo5uigB3UdCaRwYAL4gNQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

**Comment by bradzacher** [comment]
> When a.b is undefined, the original comparison produced the value true, executing the code within the if-block.

I think your wording and initial example is wrong. Please take a second to try the following code in your browser console:

```js
function test1(a) {
  if (!a || a.b === "foo") {
    return console.log(true);
  }
  console.log(false);
}

test1({});
test1({b: null});
test1({b: ""});
test1({b: "foo"});


function test2(a) {
  if (a?.b === "foo") {
    return console.log(true);
  }
  console.log(false);
}

test2({});
test2({b: null});
test2({b: ""});
test2({b: "foo"});
```

You will find that this code prints the same values for both functions.

So your statement "When a.b is undefined, the original comparison produced the value true" is incorrect.

In terms of behaviour the rule is performing exactly as expected in your example code.

`a` itself is typed to be non-nullish and so the condition `!a` always resolves to `true`. And so the rule knows that based on the types the expression can be changed to an optional chain without changing the behaviour.

[If you instead make `a` nullable then the rule no longer reports at all because it knows that the conversion would change the behaviour.](https://typescript-eslint.io/play/#ts=5.9.3&fileType=.tsx&code=CYUwxgNghgTiAEYD2A7AzgF3lAXPA3vAEYD8emMAligObwC%2B8APvCgK4QQBQXlAZvAAUAQijMWUAHRF4AXnnwARHyRJFASgJd48LvSA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0WloHsalfkwCG8WmQAWo5uigB3UdCaRwYAL4gNQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false).

**Comment by kirkwaiblinger** [comment]
Yes, the wording in the OP is incorrect, but this is still clearly a dangerous bug. _This_ rule should conservatively assume there's _some_ reason for the `!a` (namely that it might be nullable) even if it's not reflected in the types. It's the domain of no-unnecessary-condition to flag that the `!a` appears unnecessary according to the types. 

Consider code like this:
```ts
declare const a: { b?: string };

if (a == null || a.b === "foo") {
  // rule suggests to convert to `a?.b === 'foo'`, unless `a` is declared as possibly-null/undefined
  // But, `a?.b === 'foo'` is syntactically equivalent to `a != null && a.b === 'foo'`.
}
```

_Adding_ `?.` syntax, whose purpose is to handle nullable values, but with _opposite behavior in the presence of nullishness_, just because it "looks unnecessary" in the original code, is completely nonsensical and asking for trouble. The appropriate remedy to this code would be to remove the unnecessary condition/chain altogether (which is already handled by no-unnecessary-condition) not to transform it into a different, non-equivalent nullish handling strategy.

**Comment by mdm317** [comment]
>    he appropriate remedy to this code would be to remove the unnecessary condition/chain altogether (which is already handled by no-unnecessary-condition) not to transform it into a different, non-equivalent nullish handling strategy.

I agree with @kirkwaiblinger. As kirkwaiblinger pointed out, using the no-unnecessary-condition rule makes much more sensical in this situation.

**Comment by aboks** [comment]
Ah sorry, as some of you have pointed out, the wording in my report was incorrect. Instead of "When a.b is undefined ..." I meant to say "When a is undefined", but that does not make sense given the current types. It seems I stripped down my original case a bit too much to make a minimal reproducer.

My original case where I ran into this involved arrays and looked quite similar to the one provided by @Hypnosphi in https://github.com/typescript-eslint/typescript-eslint/issues/11700#issuecomment-3403114099. I think that is a better example to work from. In that example, the original expression resolves to `true` when `array[1]` is undefined, whereas the 'fixed' expression resolves to `false` in that case.

**Comment by rbourdon** [comment]
> Ah sorry, as some of you have pointed out, the wording in my report was incorrect. Instead of "When a.b is undefined ..." I meant to say "When a is undefined", but that does not make sense given the current types. It seems I stripped down my original case a bit too much to make a minimal reproducer.
> 
> My original case where I ran into this involved arrays and looked quite similar to the one provided by [@Hypnosphi](https://github.com/Hypnosphi) in [#11700 (comment)](https://github.com/typescript-eslint/typescript-eslint/issues/11700#issuecomment-3403114099). I think that is a better example to work from. In that example, the original expression resolves to `true` when `array[1]` is undefined, whereas the 'fixed' expression resolves to `false` in that case.

[#11700 (comment)](https://github.com/typescript-eslint/typescript-eslint/issues/11700#issuecomment-3403114099) suffers the same problem as the initial example, `array[1]` isn't nullable and no fix is suggested if it is made nullable:
```typescript
declare const array: ({ b?: string } | null)[];

if (!array[1] || array[1].b === "foo") {
  
}
```

**Comment by aboks** [comment]
@rbourdon But even if `array[1]` isn't nullable, it can still be `undefined` (e.g. in cases where `array.length === 1`).

**Comment by kirkwaiblinger** [comment]
Yeah, array/index access (in projects that don't use noUncheckedIndexedAccess) would be the most likely reason for an `undefined` being possible at runtime that's not reflected in the type system, even in an otherwise well-typed project. See also [limitations discussion in no-unnecessary-condition](https://typescript-eslint.io/rules/no-unnecessary-condition/#limitations).

**Comment by rbourdon** [comment]
> [@rbourdon](https://github.com/rbourdon) But even if `array[1]` isn't nullable, it can still be `undefined` (e.g. in cases where `array.length === 1`).

>Yeah, array/index access (in projects that don't use noUncheckedIndexedAccess) would be the most likely reason for an undefined being possible at runtime that's not reflected in the type system, even in an otherwise well-typed project. See also [limitations discussion in no-unnecessary-condition](https://typescript-eslint.io/rules/no-unnecessary-condition/#limitations).

Ok I see. So the array example (in projects that don't use noUncheckedIndexedAccess) and the one in the report both demonstrate code that should trigger `no-unnecessary-condition` but not this rule.

**Comment by eamodio** [comment]
I just ran into this today, as the auto-fix was dangerously changing code like:

`if (authority == null || authority.length === 0) {`

to:

``if (authority?.length === 0) {`

Which are not remotely the same check.

**Comment by kirkwaiblinger** [comment]
FYI I've also filed https://github.com/typescript-eslint/typescript-eslint/issues/11705 to downgrade autofixes to suggestions, since this would be much less harmful if it were suggesting, rather than autofixing, changes.

Related upstream issues: https://github.com/eslint/eslint/issues/18696, https://github.com/eslint/eslint/issues/20100

**Comment by mdm317** [comment]
@eamodio 
`if (authority == null || authority.length === 0) {`
This case will be fixed in this PR. [pr](https://github.com/typescript-eslint/typescript-eslint/pull/11702)

[current playground](https://typescript-eslint.io/play/#ts=5.9.2&fileType=.tsx&code=CYUwxgNghgTiAEYD2A7AzgF3lArhgFkjAJYYCeAXPJiSgOYBQxAZvABS4FGlnwC8feChwQI8AD7jseQiXIA6CCHoF%2BA%2BAAYAlPADe8AL5A&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0WloHsalfkwCG8WmQAWo5uii9o-aJHBgAviHVA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)
[pr playground](https://deploy-preview-11702--typescript-eslint.netlify.app/play/#ts=5.9.2&fileType=.ts&code=CYUwxgNghgTiAEYD2A7AzgF3lArhgFkjAJYYCe8AXPJiSgOYBQxAZvABS4FGkUC8feChwQI8AD7jseQiXIA6CCAYF4AwQAYAlPADe8AL5A&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0WloHsalfkwCG8WmQAWo5uii9o-aJHBgAviHVA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

**Comment by connorshea** [comment]
Would be great to have this fixed, the rule is misleading right now and suggests incorrect code :)

**Comment by KieranP** [comment]
Just encountered the same thing, thankfully was able to catch this during code review, otherwise it would have caused some big issues for us. Here is a very simple reproduction:

Assuming:

```js
const permissions = undefined
```

Before autofix (console.log runs):

```js
if (!permissions || permissions.length === 0) {
  console.log('Will not permit access!')
}
```

After autofix (console.log does not run!!):

```js
if (permissions?.length === 0) {
  console.log('Will not permit access!')
}
```

Huge change in functionality!

**Comment by connorshea** [comment]
Thank you!
